### PR TITLE
Implement toString ArrayList and HashMap

### PR DESCRIPTION
**Pull request by frangte**

### CODE REVIEW

File: java/lang/Integer/Integer.cpp [hunk 1]
```diff
@@ -79,6 +79,26 @@ Integer Integer::parseInt(String target) {
 }
 
 /**
+ * Compare with Integer
+ *
+ * @param o
+ * @return int
+ */
+int Integer::compareTo(const Integer &o) const {
```
Comment (dthongvl) [inline java/lang/Integer/Integer.cpp]: Please rename **o**

File: java/lang/Integer/Integer.hpp [hunk 2]
```diff
@@ -60,20 +60,8 @@ namespace Java {
...
-					if (original < o.original) {
-						return -1;
-					} else if (original == o.original) {
-						return 0;
-					}
-				}
-				return 1;
-			}
+            static Integer parseInt(String target);
+			int compareTo(const Integer &o) const;
```
Comment (dthongvl) [inline java/lang/Integer/Integer.hpp]: Please rename **o**

File: java/lang/String/StringTest.cpp [hunk 3]
```diff
@@ -85,12 +86,18 @@ TEST (JavaLang, StringCharAt) {
 	ASSERT_TRUE('H' == positionIsExist);
 
 	// Given a string - Return negative position is not exist
-	char negativePositionIsNotExist = text.charAt(-1);
-	ASSERT_TRUE(negativePositionIsNotExist == '\0');
+	try {
+		char negativePositionIsNotExist = text.charAt(-1);
+	} catch (IndexOutOfBoundsException exception) {
+		ASSERT_TRUE(true);
```
Comment (dthongvl) [inline java/lang/String/StringTest.cpp]: Is there another way to test this?

File: java/lang/String/String.cpp [hunk 4]
```diff
@@ -625,6 +659,32 @@ String String::valueOf(double target) {
...
 
 /**
+ * Returns a newly constructed string object with its value
+ * initialized to a copy of a substring of this object.
+ *
+ * @param beginIndex
+ * @return String
+ */
+String String::subString(int beginIndex) {
+   return this->subString(0, this->size);
```
Comment (dthongvl) [inline java/lang/String/String.cpp]: What **beginIndex** param is used for?
Please add test

**Review by dthongvl — CHANGES_REQUESTED** [review]

File: java/lang/String/String.cpp [hunk 5]
```diff
@@ -87,10 +107,10 @@ String String::clone() {
  * @return String
  */
 char String::charAt(int index) const{
-	if (( index < 0 ) || ( index > this->size - 1 )) {
-		return '\0';
+	if(index < 0 || index > this->size - 1) {
+		throw IndexOutOfBoundsException();
```
Comment (loint) [inline java/lang/String/String.cpp]: StringIndexOutOfBoundsException

File: java/lang/String/String.cpp [hunk 6]
```diff
@@ -156,6 +176,20 @@ Array<byte> String::getBytes() const {
...
 /**
+ * Get char to String
+ * This function use for class UUID and Long
+ *
+ * @param index
+ * @return String
+ */
+String String::getStringFromIndex(int index) {
+	if (index < 0 || index > this->size - 1) {
+		throw IndexOutOfBoundsException();
```
Comment (loint) [inline java/lang/String/String.cpp]: StringIndexOutOfBoundsException

File: java/lang/String/String.cpp [hunk 7]
```diff
@@ -447,7 +481,7 @@ boolean String::startsWith(String prefix, int toffset) const {
 	register
 #endif
 	int j = toffset;
-	
+
 	for (; index < prefix_length; index++) {
```
Comment (loint) [inline java/lang/String/String.cpp]: prefix_length => wrong coding rule

File: java/lang/String/String.cpp [hunk 8]
```diff
@@ -447,7 +481,7 @@ boolean String::startsWith(String prefix, int toffset) const {
 	register
 #endif
 	int j = toffset;
```
Comment (loint) [inline java/lang/String/String.cpp]: Please remove i,j and replace with semantic variables

File: java/util/ArrayList/ArrayList.hpp [hunk 9]
```diff
@@ -483,8 +495,24 @@ namespace Java {
...
 			 * @return string
 			 */
-			string toString() const {
-				return (string) "";
+			string toString() {
+                if(this->size() == 0) {
+                    this->backup = "[]";
+                    return this->backup.toString();
+                }
+                String mainString = "[";
```
Comment (loint) [inline java/util/ArrayList/ArrayList.hpp]: What is mainString ?

**Review by loint — CHANGES_REQUESTED** [review]

**Review by loint — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**